mio/src/lib.rs

```diff
@@ -115,77 +149,128 @@ impl TcpStream {
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    self.data = Some(data);
+                    None
+                } else {
+                    Some(Err(PollError::Other(Error::new(e, (self.offset, data)))))
                 }
             }
         }
-        let (p, c) = promise();
-        let r = self.tx.send(Message::Wait(c,
-                                           mio::EventSet::writable(),
-                                           self.tcp.clone()));
-        match r {
-            Ok(()) => {
-                let me2 = TcpStream {
-                    tcp: self.tcp.clone(),
-                    tx: self.tx.clone(),
-                };
-                p.then(move |res| {
-                    match res {
-                        Ok(()) => me2.write(offset, data),
-                        Err(e) => {
-                            futures::failed(Error::new(e, (offset, data))).boxed()
-                        }
-                    }
-                }).boxed()
-            }
-            Err(mio::channel::SendError::Io(e)) => {
-                return futures::failed(Error::new(e, (offset, data))).boxed()
-            }
-            Err(mio::channel::SendError::Disconnected(..)) => panic!("closed channel"),
+    }
+}
+
+unsafe fn slice_to_end(v: &mut Vec<u8>) -> &mut [u8] {
+    slice::from_raw_parts_mut(v.as_mut_ptr().offset(v.len() as isize),
+                              v.capacity() - v.len())
+}
+
+impl TcpStream {
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.tcp.local_addr()
+    }
+
+    pub fn peer_addr(&self) -> io::Result<SocketAddr> {
+        self.tcp.peer_addr()
+    }
+
+    pub fn ready_to_read(&self) -> MioEvent {
+        MioEvent {
+            events: mio::EventSet::readable(),
+            source: self.tcp.clone(),
+            loop_handle: self.loop_handle.clone(),
         }
     }
+
+    pub fn ready_to_write(&self) -> MioEvent {
+        MioEvent {
+            events: mio::EventSet::writable(),
+            source: self.tcp.clone(),
+            loop_handle: self.loop_handle.clone(),
+        }
+    }
+
+    // TODO: wrap in newtype
+    pub fn read(&self, mut into: Vec<u8>) -> MioFuture<ReadCompletion> {
+        self.ready_to_read().into_future(ReadCompletion {
+            tcp: self.tcp.clone(),
+            into: Some(into)
+        })
+    }
+
+    pub fn write(&self, offset: usize, data: Vec<u8>) -> MioFuture<WriteCompletion> {
+        self.ready_to_write().into_future(WriteCompletion {
+            tcp: self.tcp.clone(),
+            offset: offset,
+            data: Some(data),
+        })
+    }
+}
+
+type Waiter = Arc<Wake>;
+type Source = Arc<mio::Evented + Send + Sync>;
+
+pub struct Loop {
+    io: mio::Poll,
+    tx: mio::channel::Sender<Message>,
+    rx: mio::channel::Receiver<Message>,
+    dispatch: HashMap<usize, Waiter, BuildHasherDefault<FnvHasher>>,
```

Comment:
we probably actually want to go with a hasher here that looks like:

```rust
struct DumbHasher(usize);

impl Hasher for DumbHasher {
    fn write(&mut self, bytes: &[u8]) { panic!() }
    fn write_usize(&mut self, u: usize) { self.0 = u; }
    fn finish(&self) -> u64 { self.0 as u64 }
}
```

That should be the "fastest possible hash" for this where the keys are literally just indexes.